### DISCUSSION TITLE
PESTPP-MOU for IMS hyperparameter optimization

### DISCUSSION THREAD

**Discussion by RyanConway91** [usgs/pestpp#300] [Q&A]
I dabbling with MOU to optimize IMS. Constraint is is mass balance <0.02, objective is run time. 

Works pretty well, but the issue I am having is that the optimized IMS setting seem to be very specific to an initial condition/parameter set. When I calibrate, the model gets stuck again as IES throws different pars at it. 

I want to guassian draw of parameters as the model would see during IES included in the optimziation. I think the way to do this is via the par_stack stuff? However, I want to IMS settings to work for all members of the par_stack, not over a range of "risk". Is there a different approach that makes more sense? I thought about wrapping a simple SWP run into each MOU run, where the SWP runs the stack and MOU looks at the results of the stack rather an an individual run, but then I am losing a lot of parallelization (each SWP would be serial). 

Anyone else tried this?

**Comment by jtwhite79** [comment]
I think you are right on it - @rhugman and I have been experimenting with this as well, we are also just starting to think about using the par stack as the way to find ensemble-optimal IMS settings.  The trick will be that constraint on mass balance because running the stack will probably give you a lot of "infeasible" solutions.   You could make risk an objective - this would probably help with those infeasibilities.  We've been experimenting with making the mass balance constraint an objective, which lets you see the tradeoff you are making between solving a really tight solution vs runtime...Your idea about running swp inside of mou could also work...

**Comment by RyanConway91** [comment]
Thanks! Does look like a reasonable initial setup (not doing risk objective yet)? 
++opt_direction(min)
++mou_objectives(tot_time,tot_mb) # total run time and total mass balance error
++opt_dec_var_groups(ims_data)
++opt_constraint_groups(g_conv) #model must converge
++mou_population_schedule(mou_population_schedule.dat) #initial 100, than 20 after that (compute pool can do about 400 runs at once)
++mou_generator(de)
++opt_recalc_chance_every(1)# hyperparameter optimization highly nonlinear
++opt_chance_points(all)# hyperparameter optimization highly nonlinear
++opt_risk(0.95)
++opt_par_stack(par_stack.csv) # 20 realizations from gaussian draw of prior
++mou_save_population_every(1)

**Reply by jtwhite79** [reply]
I think thats worth a try!  let us know how it goes...
